### PR TITLE
Fix null reference exceptions on not started AsynchQueueAgent usage attempt

### PR DESCRIPTION
**Pull request by dVakulen**
After merge of https://github.com/dotnet/orleans/pull/3683 the null reference exceptions started to appear in `ProxiedMessageCenter.SendMessage` method. ([test fail 1](
https://ci.dot.net/job/dotnet_orleans/job/master/job/functional_prtest/654/consoleFull), [test fail 2](https://ci.dot.net/job/dotnet_orleans/job/master/job/functional_prtest/664/consoleFull )) 
Their cause seems to be in race condition between [adding](https://github.com/dotnet/orleans/blob/master/src/Orleans.Core/Messaging/ProxiedMessageCenter.cs#L194) of just created `GatewayConnection` (which is `AsynchQueueAgent`) to the list of connections, and [it's start](https://github.com/dotnet/orleans/blob/master/src/Orleans.Core/Messaging/ProxiedMessageCenter.cs#L273) , thus dictionary contains not ready to use connection for duration of it's `Start` method.

### CODE REVIEW

File: src/Orleans.Core/Runtime/AsynchAgent.cs [hunk 1]
```diff
@@ -122,6 +122,7 @@ public virtual void Stop()
                     {
                         State = ThreadState.StopRequested;
                         Cts.Cancel();
+                        executor = null;
```
Comment (dVakulen) [inline src/Orleans.Core/Runtime/AsynchAgent.cs]: This change is made in order to use fresh executor after Stop - Start iteration.

File: src/Orleans.Core/Runtime/AsynchAgent.cs [hunk 2]
```diff
@@ -55,6 +55,7 @@ protected AsynchAgent(string nameSuffix, ExecutorService executorService, ILogge
 
             this.executorService = executorService;
             AppDomain.CurrentDomain.DomainUnload += CurrentDomain_DomainUnload;
+            EnsureExecutorInitialized();
```
Comment (jdom) [inline src/Orleans.Core/Runtime/AsynchAgent.cs]: Is there a reason to initialize the executor before the call to Start?

Comment (dVakulen) [reply src/Orleans.Core/Runtime/AsynchAgent.cs]: Partially constructed objects aren't the nicest entities, but for this case there's indeed no reason in such initialization. Removed.

**Comment by jdom** [comment]
@dotnet-bot test functional
Reason: unrelated failure

**Comment by jdom** [comment]
Hi @dVakulen, we are still getting some non-deterministic NREs after this PR was merged, could you take a look at it, please?

```
2017-12-05T00:23:57.0422635Z System.NullReferenceException: Object reference not set to an instance of an object.
2017-12-05T00:23:57.0422635Z    at Orleans.Runtime.QueuedExecutor..ctor(String name, CancellationTokenSource cts, Boolean drainAfterCancel) in D:\build\agent\_work\15\s\src\Orleans.Core\Runtime\QueuedExecutor.cs:line 34
2017-12-05T00:23:57.0422635Z    at Orleans.Runtime.ExecutorService.GetExecutor(GetExecutorRequest request) in D:\build\agent\_work\15\s\src\Orleans.Core\Runtime\ExecutorService.cs:line 29
2017-12-05T00:23:57.0422635Z    at Orleans.Runtime.AsynchAgent.EnsureExecutorInitialized() in D:\build\agent\_work\15\s\src\Orleans.Core\Runtime\AsynchAgent.cs:line 171
2017-12-05T00:23:57.0422635Z    at Orleans.Runtime.AsynchAgent.Start() in D:\build\agent\_work\15\s\src\Orleans.Core\Runtime\AsynchAgent.cs:line 103
2017-12-05T00:23:57.0422635Z    at Orleans.Messaging.GatewayConnection.Start() in D:\build\agent\_work\15\s\src\Orleans.Core\Messaging\GatewayConnection.cs:line 77
2017-12-05T00:23:57.0422635Z    at Orleans.Messaging.ProxiedMessageCenter.SendMessage(Message msg) in D:\build\agent\_work\15\s\src\Orleans.Core\Messaging\ProxiedMessageCenter.cs:line 273
2017-12-05T00:23:57.0422635Z    at Orleans.Runtime.QueuedExecutor.RunNonBatching() in D:\build\agent\_work\15\s\src\Orleans.Core\Runtime\QueuedExecutor.cs:line 114
2017-12-05T00:23:57.0422635Z    at Orleans.Runtime.QueuedExecutor.<.ctor>b__4_1(Object _) in D:\build\agent\_work\15\s\src\Orleans.Core\Runtime\QueuedExecutor.cs:line 40
2017-12-05T00:23:57.0422635Z    at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
2017-12-05T00:23:57.0422635Z    at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
2017-12-05T00:23:57.0422635Z    at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
2017-12-05T00:23:57.0422635Z    at System.Threading.ThreadHelper.ThreadStart()
```

**Pull request merged**